dev/dev_setup.sh

```diff
@@ -0,0 +1,306 @@
+                $MACOS12 | $MACOS14)
+                    notice "Installing unzip via Homebrew"
+                    logged brew install unzip && USED_BREW=true;;
+                $UBUNTU20 | $UBUNTU22)
+                    notice "Installing unzip via apt-get"
+                    logged apt-get install -y unzip;;
+                *) unreachable "Unsupported platform; did not install unzip"; return;;
+            esac
+        fi
+        logged unzip solvers.bin.zip
+
+        # If we want to install more solvers by default, we can do so here
+        for solver in cvc4 cvc5 z3; do
+            if ! is_installed $solver ; then
+                notice "Installing $solver"
+                chmod u+x $solver
+                # Try installing without sudo first, because some environments
+                # don't have it installed by default
+                mv $solver /usr/local/bin || sudo mv $solver /usr/local/bin
+            fi
+        done
+
+        cd $HERE
+        rm -r $solvers_dir
+    else
+        notice "Not installing cvc4, cvc5, or z3 solvers because they already exist"
+    fi
+
+    # Make sure the installed versions are the versions that have been tested
+    check_version cvc4 $WHAT4_CVC4_VERSION
+    check_version cvc5 $WHAT4_CVC5_VERSION
+    check_version z3 $WHAT4_Z3_VERSION
+}
+
+# Checks the version of the given command
+# Usage: check_version <command> <expected version>
+#
+# Assumes that the command can be called with `--version`
+check_version() {
+    cmd="$1"
+    expected="$@"
+    version_file=$(mktemp)
+    $cmd --version > $version_file
+    if ! (grep -iq "$expected" $version_file); then
+        actual_version=$(grep -i -m 1 "version" $version_file)
+        notice "Your version of $cmd is unexpected; expected $expected"
+        notice "    Got: $actual_version"
+        notice "    To ensure compatibility, you might want to uninstall the "\
+            "existing version and re-run this script."
+    fi
+    rm $version_file
+}
+
+
+put_brew_in_path() {
+    if $USED_BREW; then
+        # `brew --prefix` is different on macOS 12 and macOS 14
+        echo "export CPATH=$(brew --prefix)/include" >> $VAR_FILE
+        echo "export LIBRARY_PATH=$(brew --prefix)/lib" >> $VAR_FILE
+    fi
+}
+
+set_ubuntu_language_encoding() {
+    case $CRYPTOL_PLATFORM in
+        $UBUNTU20 | $UBUNTU22)
+            notice "Adding language environment variables to $VAR_FILE"
+            echo "export LANG=C.UTF-8" >> $VAR_FILE
+            echo "export LC_ALL=C.UTF-8" >> $VAR_FILE;;
+        *) ;;
+    esac
+}
+
+# Make sure script is running on a supported platform
+notice "Checking whether platform is supported"
+CRYPTOL_PLATFORM=$(supported_platform)
+if [ -z "$CRYPTOL_PLATFORM" ]; then
+    echo "Unsupported platform; this script supports Ubuntu 20.04, Ubuntu 22.04, and macOS 14"
+    exit 1
+fi
+
+update_submodules
+install_ghcup
+install_gmp
+install_zlib
+install_solvers
+put_brew_in_path
+set_ubuntu_language_encoding
```

Comment:
I appreciate the way this structure summarizes the set of operations being performed.  It is a good "self-documenting" summary of what the code is doing for both the user of this script or someone working in an environment this script doesn't support and who wants to know what it does.  To support both of these, I'd recommend adding a reference to this portion in the preface comments; for example, add a "# Main operations" comment just above line 297 and then finish the paragraph in lines 10-13 with "For a list of the setup operations performed by this script, see the "Main Operations" section below (near the end of this file)."

Reply:
This is nice, I added it.